src/Build/Collections/PropertyDictionary.cs

```diff
@@ -529,17 +529,25 @@ internal Dictionary<string, string> ToDictionary()
             }
         }
 
-        internal void Enumerate(Action<string, string> keyValueCallback)
+        internal IEnumerable<(string propertyName, string propertyValue)> Enumerate()
         {
             lock (_properties)
             {
                 foreach (var kvp in (ICollection<T>)_properties)
                 {
-                    keyValueCallback(kvp.Key, EscapingUtilities.UnescapeAll(kvp.EscapedValue));
+                    yield return (kvp.Key, EscapingUtilities.UnescapeAll(kvp.EscapedValue));
                 }
             }
         }
 
+        internal void Enumerate(Action<string, string> keyValueCallback)
```

Comment:
are you creating a custom observer pattern? if so, consider the reactive extensions

Reply:
I do not thing Rx.NET would be a good investment choice for msbuild at this point of time - but since this is being brought up repetitively:
* https://github.com/dotnet/msbuild/pull/10768#discussion_r1793369957
* https://github.com/dotnet/msbuild/pull/9132#discussion_r1303075800

Let's discuss offline what it would mean (extra binary dependency, changes to public API, rewriting subscribing model to observing model) and what would be the  added value in return.